File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
def solution(numbers):
    answer = [-1 for _ in range(len(numbers))]
    stack = []
    
    for i, num in enumerate(numbers):
        while stack and stack[-1][1] < num:
            idx, _ = stack.pop()
            answer[idx] = num         
                
        stack.append([i, num])
    return answer

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.Deque;
import java.util.ArrayDeque;


class Solution {
    public int[] solution(int[] numbers) {
        int[] answer = new int[numbers.length];
        Deque<Integer> dq = new ArrayDeque<Integer>();
        
        for (int i=numbers.length - 1; i >= 0; i --){
            int num = numbers[i];
            while (!dq.isEmpty() && num >= dq.getFirst()) {
                dq.poll();
            }
            if (!dq.isEmpty()){
                answer[i] = dq.getFirst();
            } else {
                answer[i] = -1;
            }
            dq.addFirst(num);
        }
        return answer;
    }
}